src/expr/sygus_grammar.cpp

```diff
@@ -0,0 +1,175 @@
+#include "printer/smt2/smt2_printer.h"
+
+namespace cvc5::internal {
+
+SygusGrammar::SygusGrammar(const std::vector<Node>& sygusVars,
+                           const std::vector<Node>& ntSyms)
+    : d_sygusVars(sygusVars), d_ntSyms(ntSyms)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  for (const Node& ntSym : ntSyms)
+  {
+    d_sdts.emplace(ntSym, SygusDatatype(ntSym.getName()));
+    d_ntsToUnres.emplace(ntSym, nm->mkUnresolvedDatatypeSort(ntSym.getName()));
+  }
+}
+
+bool isId(const Node& n)
+{
+  return n.getKind() == kind::LAMBDA && n[0].getNumChildren() == 1
+         && n[0][0] == n[1];
+}
+
+void SygusGrammar::addRule(const Node& ntSym, const Node& rule)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  std::vector<Node> args;
+  std::vector<TypeNode> cargs;
+  Node op = purifySygusGNode(rule, args, cargs);
+  std::stringstream ss;
+  ss << op.getKind();
+  if (!args.empty())
+  {
+    Node lbvl = nm->mkNode(kind::BOUND_VAR_LIST, args);
+    op = nm->mkNode(kind::LAMBDA, lbvl, op);
+  }
+  // assign identity rules a weight of 0.
+  d_sdts.at(ntSym).addConstructor(op, ss.str(), cargs, isId(op) ? 0 : -1);
+}
+
+void SygusGrammar::addRules(const Node& ntSym, const std::vector<Node>& rules)
+{
+  for (const Node& rule : rules)
+  {
+    addRule(ntSym, rule);
+  }
+}
+
+void SygusGrammar::addAnyConstant(const Node& ntSym, const TypeNode& tn)
+{
+  d_allowConst.emplace(ntSym);
+}
+
+void SygusGrammar::addAnyVariable(const Node& ntSym)
+{
+  // each variable of appropriate type becomes a sygus constructor in sdt.
+  for (const Node& v : d_sygusVars)
+  {
+    if (v.getType() == ntSym.getType())
+    {
+      d_sdts.at(ntSym).addConstructor(v, v.getName(), {});
+    }
+  }
+}
+
+TypeNode SygusGrammar::resolve()
+{
+  if (!isResolved())
+  {
+    NodeManager* nm = NodeManager::currentNM();
+    Node bvl;
+    if (!d_sygusVars.empty())
+    {
+      bvl = nm->mkNode(kind::BOUND_VAR_LIST, d_sygusVars);
+    }
+    std::vector<DType> datatypes;
+    for (const Node& ntSym : d_ntSyms)
+    {
+      bool allowConst = d_allowConst.find(ntSym) != d_allowConst.cend();
+      d_sdts.at(ntSym).initializeDatatype(
+          ntSym.getType(), bvl, allowConst, false);
+      datatypes.push_back(d_sdts.at(ntSym).getDatatype());
+    }
+    d_datatype = nm->mkMutualDatatypeTypes(datatypes)[0];
+  }
+  // return the first datatype
+  return d_datatype;
+}
+
+bool SygusGrammar::isResolved()
+{
+  return !d_datatype.isNull();
+}
+
+const std::vector<Node>& SygusGrammar::getSygusVars() const
+{
+  return d_sygusVars;
+}
+
+const std::vector<Node>& SygusGrammar::getNtSyms() const { return d_ntSyms; }
+
+size_t SygusGrammar::getNumConstructors(Node ntSym) const
+{
+  return d_sdts.at(ntSym).getNumConstructors();
+}
+
+std::string SygusGrammar::toString() const
+{
+  std::stringstream ss;
+  // clone this grammar before printing it to avoid freezing it.
+  return static_cast<printer::smt2::Smt2Printer*>(Printer::getPrinter(ss))
```

Comment:
You don't need to access the printer, I believe you can just invoke the static method here, i.e. `print::smt2::Smt2Printer::sygusGrammarString(..)`